### ISSUE TITLE
SameDiff: multiple concurrent execution streams (single GPU)

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8385]
In many neural networks, we have multiple operations that can be executed concurrently. A well known example is inception blocks:
![image](https://user-images.githubusercontent.com/2360237/68671865-78578b00-05a4-11ea-8926-447bb4145c87.png)

Here, there are 4 operations that can be executed in parallel.

Due to CUDA launch overheads and the presence of small operations, it will likely be faster to execute multiple operations concurrently even on a single GPU (compared to running one operation at a time).

The new execution framework (InferenceSession etc) makes this easy - we already have a queue of operations that are ready to be executed, so it should be a case of adding threads on top of the existing structure.
There may be some implementation overlap with multi-GPU execution also.

**Comment by agibsonccc** [comment]
Closing — the CUDA backend has been completely rewritten with new memory management, DSP execution engine, CUDA graph replay, and Triton compilation. If you still experience GPU-related issues with the upcoming release, please open a new issue.
